File: problems/data-structures-n-algorithms/hard/76.js
// 76. Minimum Window Substring
// https://leetcode.com/problems/minimum-window-substring/

const copyMap = (o, d) => {
	for (const key in o) {
		d[key] = o[key];
	}
};

// time limit exceeded for last test case
const minWindow = function (s, t) {
	const map = {};
	const cpMap = {};
	for (const item of t) {
		if (!map[item]) {
			map[item] = 1;
		} else {
			map[item]++;
		}
	}
	copyMap(map, cpMap);

	let minFirst = -1;
	let minLast = -1;
	let first = -1;
	let second = -1;
	let last = -1;
	let len = 0;

	for (let i = 0; i < s.length; i++) {
		const c = s[i];
		if (cpMap[c] && first === -1) {
			first = i;
			cpMap[c]--;
			len++;
		} else if (cpMap[c]) {
			if (len === 1 && second === -1) {
				second = i;
			}
			cpMap[c]--;
			len++;
		}

		if (first !== -1 && second === -1 && s[i] === s[first] && i > first) {
			second = i;
		}

		if (len === t.length) {
			last = i;
			if (second !== -1) {
				i = second - 1;
			}
			second = -1;
			copyMap(map, cpMap);
			len = 0;

			if (minLast - minFirst > last - first || minLast === -1) {
				minLast = last;
				minFirst = first;
			}
			last = -1;
			first = -1;
		}
	}

	return s.substring(minFirst, minLast + 1);
};

function main() {
	debugger;
	const res = minWindow("ADOBECODEBANC", "ABC");
	// const res = minWindow("a", "a");
	// const res = minWindow("bba", "ab");
	console.log(res);
}

main();

File: problems/data-structures-n-algorithms/hard/76.cpp
// 76. Minimum Window Substring
// https://leetcode.com/problems/minimum-window-substring/

#include <deque>
#include <iostream>
#include <limits.h>
#include <queue>
#include <unordered_map>
#include <vector>

using namespace std;

string minWindow(string s, string t) {
	int map[128] = {0};
	int end = 0, begin = 0, head = 0, counter = t.size(), d = INT_MAX;

	for (char c : t)
		map[c]++;

	while (end < s.size()) {
		if (map[s[end++]]-- > 0) {
			counter--;
		}

		while (counter == 0) {
			if (end - begin < d)
				d = end - (head = begin);

			if (map[s[begin++]]++ == 0)
				counter++;
		}
	}

	return s.substr(head, d == INT_MAX ? 0 : d);
}

int main() {
	// string s = "ADOBECODEBANC";
	// string t = "ABC";
	// string s = "a";
	// string t = "a";
	// string s = "bba";
	// string t = "ab";
	string s = "a";
	string t = "aa";
	// string s = "aa";
	// string t = "aa";

	string res = minWindow(s, t);

	cout << res << endl;
}